tests/kernel/mem_protect/userspace/src/main.c

```diff
@@ -0,0 +1,231 @@
+ */
+#ifndef CONFIG_ARM
+FUNC_NORETURN
+#endif
+void _SysFatalErrorHandler(unsigned int reason, const NANO_ESF *pEsf)
+{
+	INFO("Caught system error -- reason %d\n", reason);
+	ztest_test_pass();
+#ifndef CONFIG_ARM
+	CODE_UNREACHABLE;
+#endif
+}
+
+static void is_usermode(void)
+{
+	/* Confirm that we are in fact running in user mode. */
+	zassert_true(_is_user_context(), "thread left in kernel mode\n");
+}
+
+static void priv_insn(void)
+{
+	/* Try to invoke what should require a privileged instruction. */
+	unsigned int key = irq_lock();
+
+	irq_unlock(key); /* just in case we succeeded */
+	zassert_unreachable("irq_lock did not fault\n");
+}
+
+static void write_control(void)
+{
+	/* Try to write to a control register. */
+#if defined(CONFIG_X86)
+	__asm__ volatile (
+		"mov %cr0, %eax;\n\t"
+		"and $0xfffeffff, %eax;\n\t"
+		"mov %eax, %cr0;\n\t"
+		);
+#elif defined(CONFIG_ARM)
+	__asm__ volatile (
+		"mrs r2, CONTROL;\n\t"
+		"bic r2, #1;\n\t"
+		"msr CONTROL, r2;\n\t"
+		);
+#endif
+	zassert_unreachable("Write to control register did not fault\n");
+}
+
+static void disable_mmu_mpu(void)
+{
+	/* Try to disable memory protections. */
+#if defined(CONFIG_X86)
+	__asm__ volatile (
+		"mov %cr0, %eax;\n\t"
+		"and $0x7ffeffff, %eax;\n\t"
+		"mov %eax, %cr0;\n\t"
+		);
+#elif defined(CONFIG_ARM)
+	arm_core_mpu_disable();
+#endif
+	zassert_unreachable("Disable MMU/MPU did not fault\n");
+}
+
+static void read_kernram(void)
+{
+	/* Try to read from kernel RAM. */
+	void *p = _current->init_data;
+
+	printk("%p\n", p);
+	zassert_unreachable("Read from kernel RAM did not fault\n");
+}
+
+static void write_kernram(void)
+{
+	/* Try to write to kernel RAM. */
+	_current->init_data = NULL;
+	zassert_unreachable("Write to kernel RAM did not fault\n");
+}
+
+extern int _k_neg_eagain;
+
+static void write_kernro(void)
+{
+	/* Try to write to kernel RO. */
+	_k_neg_eagain = -EINVAL;
+	zassert_unreachable("Write to kernel RO did not fault\n");
+}
+
+static void write_kerntext(void)
+{
+	/* Try to write to kernel text. */
+	memcpy(&_is_thread_essential, 0, 4);
+	zassert_unreachable("Write to kernel text did not fault\n");
+}
+
+__kernel static int kernel_data;
+
+static void read_kernel_data(void)
+{
+	/* Try to read from embedded kernel data. */
+	int value = kernel_data;
+
+	printk("%d\n", value);
+	zassert_unreachable("Read from __kernel data did not fault\n");
+}
+
+static void write_kernel_data(void)
+{
+	kernel_data = 1;
+	zassert_unreachable("Write to  __kernel data did not fault\n");
+}
+
+/*
+ * volatile to avoid compiler mischief.
+ */
+volatile int *ptr = NULL;
+#if defined(CONFIG_X86)
+volatile size_t size = MMU_PAGE_SIZE;
+#elif defined(CONFIG_ARM) && defined(CONFIG_PRIVILEGED_STACK_SIZE)
+volatile size_t size = CONFIG_PRIVILEGED_STACK_SIZE;
+#else
+volatile size_t size = 512;
+#endif
+
+static void read_kernel_stack(void)
+{
+	/* Try to read from kernel stack. */
+	int s[1];
+
+	s[0] = 0;
+	ptr = &s[0];
+	ptr -= size;
```

Comment:
This will move ptr back by size * sizeof(int). Is this correct? Or this should be `ptr = (unsigned char *)ptr - size` instead?

Reply:
Sorry, my mistake.  Fixed.